File: Day 13/Python/sol.py
"""
FILE: Day 13/Python/sol.py
Author: Alex Jones
Desc: Solution to day 13 problems (25 & 26) for Advent of Code 2024, solved in Python 3.
"""
data = open("Day 13/data.txt", "r").read().strip().replace("\r","").split("\n\n")

"""There is actually only 1 way to solve each - use simultaneous eqs.
[1] ax * A + bx * B = gx
[2] ay * A + by * B = gy
Re-arrange [2]: A = (gy - by * B)/ay
Sub in [1] to get [3]: (ax * (gy - by * B)) / ay + bx * B = gx
Re-arrange [3] to get [4]: (ax * gy)/ay - (ax * by * B))/ay + bx * B = gx
                            bx * B - (ax * by * B)/ay = gx - (ax * gy)/ay
                            B(bx - (ax * by)/ay) = gx - (ax * gy)/ay
                            B = (gx - (ax * gy)/ay)/(bx - (ax * by)/ay)
                            B = (gx * ay - ax * gy)/(ay * (bx - (ax * by)/ay))
                            B = (gx * ay - ax * gy)//(ay * bx - ax * by)
 Use [4] and then [2] to solve for B and A."""
def least_moves(moves: list[tuple[int,int]], goal: tuple[int,int]) -> int:
    (ax, ay), (bx, by), (gx, gy) = moves[0], moves[1], goal
    numerator, denom = gx * ay - ax * gy, ay * bx - ax * by
    if denom == 0: return 0  # Impossible to solve
    elif numerator % denom != 0: return 0  # Non-integer, impossible in context
    B = numerator // denom
    numerator, denom = gy - by * B, ay
    if numerator % denom != 0: return 0  # Non-integer, impossible in context
    A = numerator // denom
    return A * 3 + B
    

p1, p2 = 0, 0
for entry in data:
    entry = [row.split(":")[-1].split(",") for row in entry.split("\n")]
    entry = [[int(d.replace("=","+").split("+")[-1].strip()) for d in row]
             for row in entry]
    p1 += least_moves(entry[:-1], entry[-1])
    p2 += least_moves(entry[:-1], [v+10000000000000 for v in entry[-1]])

print("Problem 25:", p1)
print("Problem 26:", p2)

File: Day 13/Rust/sol.rs
/*
 * FILE: Day 13/Rust/sol.rs
 * Author: Alex Jones
 * Desc: Solution to day 13 problems (25 & 26) for Advent of Code 2024, solved in Rust.
 */
use itertools::Itertools;
use std::fs;

fn least_moves(moves: &[(i64, i64); 2], goal: (i64, i64)) -> i64 {
    let (ax, ay) = moves[0];
    let (bx, by) = moves[1];
    let (gx, gy) = goal;
    let numerator = gx * ay - ax * gy;
    let denominator = ay * bx - ax * by;
    if denominator == 0 || numerator % denominator != 0 {
        return 0; // Impossible to solve or non-integer solution (impossible)
    }
    let b_mul = numerator / denominator;
    let numerator = gy - by * b_mul;
    let denominator = ay;
    if numerator % denominator != 0 {
        return 0; // Non-integer solution (impossible)
    }
    let a_mul = numerator / denominator;
    a_mul * 3 + b_mul
}

pub fn solve() {
    let contents: String = fs::read_to_string("Day 13/data.txt")
        .expect("Could not read data file")
        .trim()
        .replace("\r", "");
    let data: Vec<[(i64, i64); 3]> = contents
        .split("\n\n")
        .map(|entry| {
            entry
                .split('\n')
                .map(|row| {
                    row.split(':')
                        .last()
                        .unwrap()
                        .split(',')
                        .map(|d| {
                            d.split(['+', '='])
                                .last()
                                .unwrap()
                                .trim()
                                .parse::<i64>()
                                .unwrap()
                        })
                        .collect_tuple()
                        .unwrap()
                })
                .collect::<Vec<_>>()
                .try_into()
                .unwrap()
        })
        .collect();

    println!(
        "Problem 25: {}",
        data.iter()
            .map(|entry| least_moves(&entry[0..2].try_into().unwrap(), entry[2]))
            .sum::<i64>()
    );
    const OFFSET: i64 = 10000000000000;
    println!(
        "Problem 26: {}",
        data.iter()
            .map(|entry| least_moves(
                &entry[0..2].try_into().unwrap(),
                (entry[2].0 + OFFSET, entry[2].1 + OFFSET)
            ))
            .sum::<i64>()
    );
}
